src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -4035,10 +4037,18 @@ private void ParseParameterModifiers(SyntaxListBuilder modifiers)
                 {
                     case SyntaxKind.ThisKeyword:
                         modifier = CheckFeatureAvailability(modifier, MessageID.IDS_FeatureExtensionMethod);
+                        if (seenRefModifier)
+                        {
+                            modifier = CheckFeatureAvailability(modifier, MessageID.IDS_FeatureRefExtensionMethods);
```

Comment:
Another thought (not blocking this PR): My approach for C# 7.1 features was to not rely on the generic diagnostic from CheckFeatureAvailability, but to instead use diagnostic messages customized to each situation, but including "Please use version X or above." at the end.
The advantage is that the message is tailored and more user-friendly. The downside is that you need to add that new error code to the UpgradeProject code fixer.
I don't mind if you prefer to rely on CheckFeatureAvailability, but I wanted to bring this up.

---
In reply to: [123613868](https://github.com/dotnet/roslyn/pull/20403#discussion_r123613868) [](ancestors = 123613868)